### ISSUE TITLE
Error from read_feather (in the feather R library)

### ISSUE THREAD

**Issue by arturosaco** [wesm/feather#150]
I've been getting the following error when using read_feather():

```
 *** caught segfault ***
address 0x7f9b75bab010, cause 'memory not mapped'

Traceback:
 1: .Call("feather_coldataFeather", PACKAGE = "feather", feather,     indexes)
 2: coldataFeather(x, i)
 3: `[.feather`(x, )
 4: x[]
 5: as_data_frame.feather(data)
 6: as_data_frame(data)
 7: read_feather(file.match)
```

My sessionInfo():

```
R version 3.2.1 (2015-06-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.3 LTS

locale:
 [1] LC_CTYPE=en_US.UTF-8          LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8           LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8       LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8          LC_NAME=en_US.UTF-8
 [9] LC_ADDRESS=en_US.UTF-8        LC_TELEPHONE=en_US.UTF-8
[11] LC_MEASUREMENT=en_US.UTF-8    LC_IDENTIFICATION=en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
 [1] feather_0.0.1         metricsgraphics_0.8.5 dygraphs_0.5
 [4] RColorBrewer_1.1-2    htmlwidgets_0.5       htmltools_0.2.6
 [7] DT_0.1                shinydashboard_0.5.1  shiny_0.12.2
[10] lubridate_1.3.3       helpRs_0.1            magrittr_1.5
[13] ggplot2_2.1.0         RPostgres_0.1-1       data.table_1.9.6
[16] dplyr_0.4.3


loaded via a namespace (and not attached):
 [1] Rcpp_0.12.5          futile.logger_1.4.1  plyr_1.8.1
 [4] R.methodsS3_1.7.1    futile.options_1.0.0 R.utils_2.3.0
 [7] tools_3.2.1          digest_0.6.8         packrat_0.4.1-1
[10] tibble_1.0           lattice_0.20-31      memoise_0.2.1
[13] gtable_0.1.2         DBI_0.4-1            parallel_3.2.1
[16] rJava_0.9-8          mailR_0.4.1          stringr_1.0.0
[19] httr_1.1.0           grid_3.2.1           R6_2.1.2
[22] lambda.r_1.1.7       scales_0.4.0         urltools_1.4.0
[25] tldextract_1.0       assertthat_0.1       mime_0.4
[28] colorspace_1.2-4     xtable_1.7-4         httpuv_1.3.3
[31] stringi_1.0-1        munsell_0.4.2        chron_2.3-47
[34] zoo_1.7-11           R.oo_1.20.0
```

And the `str(the.table_I_tried.to.save)` of the object I'm writing:

```
Classes ‘data.table’ and 'data.frame':  1888384 obs. of  5 variables:
 $ uuid         : chr  "00105052-ff64-43ed-9405-2d4bf847ac40" "0001a6a0-2cf2-49ca-adc3-ff0cd87cf1aa" "00027a78-6774-48cc-a53f-6c985321d24f" "0000867e-666e-4c99-9bc0-1a153bd5e53e" ...
 $ status          : chr  "no data" "success" "no data" "success" ...
 $ update_date     : chr  "2016-02-15" "2016-02-15" "2016-02-15" "2016-02-15" ...
 $ table_name      : chr  "location" "location" "location" "location" ...
 $ last_update_date: chr  "2016-02-15" "2016-02-15" "2016-02-15" "2016-02-15" ...
 - attr(*, ".internal.selfref")=<externalptr>
```

I wish I could produce a reproducible example but I haven't been able to come up with one, and unfortunately I can't post the data I'm using. Fwiw, I tried writing a plain data.frame instead of a data.table but the same happened. Afaik there shouldn't be anything funky in the values of the variables. 

A couple more details I've noticed:

- This tends to happen with large data.frames
- If I first read the data.frame from an .rds file and then read the same object from a .feather file this error doesn't happen (observed this while debugging).

**Comment by dtmngngnnj** [comment]
I'm getting the same error:
```
    Traceback:
     1: .Call("feather_coldataFeather", PACKAGE = "feather", feather,     indexes)
     2: coldataFeather(x, i)
     3: `[.feather`(x, )
     4: x[]
     5: as_data_frame.feather(data)
     6: as_data_frame(data)
     7: read_feather(filename)
```
and found out that it's due to the size of the data: reading a smaller version of the same file gives no problem. 

### Reproduce the error

Download the yellow cab trip sheet data for January 2015 from [www.nyc.gov/html/tlc/html/about/trip_record_data.shtml](http://www.nyc.gov/html/tlc/html/about/trip_record_data.shtml), which is about 1.9G in size.

Read csv file and write feather file like this in R:

    library(feather)
    csv_file='yellow_tripdata_2015-01.csv'
    feather_file='yellow_tripdata_2015-01.feather'
    df<-read.table(csv_file,header=TRUE,sep=",",quote="",stringsAsFactors=F, na.strings = "")
    write_feather(df, feather_file)

If you try to read the feather file again in the same R-session, it will work, so exit your R-session and startup a new one, in which you execute:

    library(feather)
    feather_file='yellow_tripdata_2015-01.feather'
    df <- read_feather(feather_file)

Output:

     *** caught segfault ***
    address 0x7f3c405ff010, cause 'memory not mapped'

    Traceback:
     1: .Call("feather_coldataFeather", PACKAGE = "feather", feather,     indexes)
     2: coldataFeather(x, i)
     3: `[.feather`(x, )
     4: x[]
     5: as_data_frame.feather(data)
     6: as_data_frame(data)
     7: read_feather(feather_file)

Trying it for various file-sizes, I start getting the error as of size 78M. 

There's more detail to my findings over here:
http://data.munging.ninja/pubcoms/20160526_bug_feather_mem_not_mapped/

**Comment by wesm** [comment]
@dtmngngnnj interesting, so you're getting this failure on a feather file that's less than 2GB? May be a different bug than #155

**Comment by dtmngngnnj** [comment]
After testing with a subset of columns, I strongly suspect it is 'chr' related!

This is the structure of the dataframe:

```
> str(df)
Classes ‘tbl_df’, ‘tbl’ and 'data.frame':   10000 obs. of  19 variables:
 $ VendorID             : int  2 1 1 1 1 1 1 1 1 1 ...
 $ tpep_pickup_datetime : chr  "2015-01-15 19:05:39" "2015-01-10 20:33:38" ..
 $ tpep_dropoff_datetime: chr  "2015-01-15 19:23:42" "2015-01-10 20:53:28" ..
 $ passenger_count      : int  1 1 1 1 1 1 1 3 3 2 ...
 $ trip_distance        : num  1.59 3.3 1.8 0.5 3 9 2.2 0.8 18.2 0.9 ...
 $ pickup_longitude     : num  -74 -74 -74 -74 -74 ...
 $ pickup_latitude      : num  40.8 40.7 40.8 40.7 40.8 ...
 $ RateCodeID           : int  1 1 1 1 1 1 1 1 2 1 ...
 $ store_and_fwd_flag   : chr  "N" "N" "N" "N" ...
 $ dropoff_longitude    : num  -74 -74 -74 -74 -74 ...
 $ dropoff_latitude     : num  40.8 40.8 40.8 40.7 40.7 ...
 $ payment_type         : int  1 1 2 2 2 1 2 1 2 1 ...
 $ fare_amount          : num  12 14.5 9.5 3.5 15 27 14 7 52 6.5 ...
 $ extra                : num  1 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0 0.5 ...
 $ mta_tax              : num  0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 ...
 $ tip_amount           : num  3.25 2 0 0 0 6.7 0 1.66 0 1.55 ...
 $ tolls_amount         : num  0 0 0 0 0 5.33 0 0 5.33 0 ...
 $ improvement_surcharge: num  0.3 0.3 0.3 0.3 0.3 0.3 0.3 0.3 0.3 0.3 ...
 $ total_amount         : num  17.1 17.8 10.8 4.8 16.3 ...
```

If I retain only the **num** and **int** columns (when converting CSV -> feather), there is no issue when reading back the feather file in a new R session. But when I include a **chr** column, I get the error. 

Here an overview of the cases I tested up to now:

```
df=df[,c('VendorID','tip_amount')]
--> OK

df=df[,c('VendorID','tip_amount','tpep_pickup_datetime')]
--> ERROR

df=df[,c('VendorID' ,'passenger_count' ,'trip_distance' ,'pickup_longitude' ,
'pickup_latitude' ,'RateCodeID' ,'dropoff_longitude' ,'dropoff_latitude' ,
'payment_type' ,'fare_amount' ,'extra' ,'mta_tax' ,'tip_amount' ,
'tolls_amount' ,'improvement_surcharge' ,'total_amount')]
--> OK

df=df[,c('tip_amount','tpep_dropoff_datetime')]
--> ERROR
```

I'm doing some further testing (maybe it's data related).

**Comment by dtmngngnnj** [comment]
@wesm this is the output of gdb when reproducing this issue: 
```
#0  0x00007ffff78f3d86 in SET_STRING_ELT () from /usr/lib/R/lib/libR.so
#1  0x00007ffff2a95ed9 in toSEXP (val=val@entry=0x22e1860) at feather-types.cpp:136
#2  0x00007ffff2a96958 in toSEXP (x=std::unique_ptr<feather::Column> containing 0x22e1840) at feather-types.cpp:222
#3  0x00007ffff2a949f9 in coldataFeather (feather=..., indexes=...) at feather-read.cpp:102
#4  0x00007ffff2a90afc in feather_coldataFeather (featherSEXP=<optimized out>, indexesSEXP=<optimized out>) at RcppExports.cpp:50
#5  0x00007ffff788ccc8 in ?? () from /usr/lib/R/lib/libR.so
#6  0x00007ffff78c8548 in Rf_eval () from /usr/lib/R/lib/libR.so
#7  0x00007ffff78ca2c8 in ?? () from /usr/lib/R/lib/libR.so
..
..
```

**Comment by dtmngngnnj** [comment]
Hoping the problem would be solved with a more recent version of the toolchain, I tested feather on debian-unstable (aka sid),  **alas the problem lives on** : 

```
> library(feather)
>  df <- read_feather('yellow_tripdata_2015-01-500001.feather')

 *** caught segfault ***
address 0x7f3d435b6010, cause 'memory not mapped'

Traceback:
 1: .Call("feather_coldataFeather", PACKAGE = "feather", feather,     indexes)
 2: coldataFeather(x, i)
 3: `[.feather`(x, )
 4: x[]
 5: as_data_frame.feather(data)
 6: as_data_frame(data)
 7: read_feather("yellow_tripdata_2015-01-500001.feather")
```


Version-info:

```
debian: /etc/debian_version -> "stretch/sid"

g++ -v : -> "gcc version 5.3.1 20160528 (Debian 5.3.1-21)" 

R:
    platform       x86_64-pc-linux-gnu
    major          3
    minor          3.0
    year           2016
    month          05
    day            03
    svn rev        70573
    language       R
    version.string R version 3.3.0 (2016-05-03)
```

**Comment by wesm** [comment]
Can you run with valgrind? (`valgrind --tool=memcheck R`) perhaps there is memory corruption

**Comment by hadley** [comment]
I think the preferred way to run valgrind with R is `R --debugger="valgrind"`

**Comment by dtmngngnnj** [comment]
Here you go..

``` 
R  --debugger="valgrind"
==3335== Memcheck, a memory error detector
==3335== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==3335== Using Valgrind-3.10.0 and LibVEX; rerun with -h for copyright info
==3335== Command: /usr/lib/R/bin/exec/R
==3335== 

R version 3.3.0 (2016-05-03) -- "Supposedly Educational"
Copyright (C) 2016 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(feather)
> df <- read_feather('yellow_tripdata_2015-01-500001.feather')
==3335== Invalid read of size 1
==3335==    at 0x4F72D86: SET_STRING_ELT (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7ED8: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:136)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4CE0F: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4712E: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4914D: ??? (in /usr/lib/R/lib/libR.so)
==3335==  Address 0x124d4040 is 0 bytes inside a block of size 4,000,040 free'd
==3335==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==3335==    by 0x4F7694C: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F78BAB: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F28DAC: Rf_mkCharLenCE (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7EC6: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:135)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335== 
==3335== Invalid read of size 4
==3335==    at 0x4F72DAE: SET_STRING_ELT (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7ED8: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:136)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4CE0F: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4712E: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4914D: ??? (in /usr/lib/R/lib/libR.so)
==3335==  Address 0x124d4060 is 32 bytes inside a block of size 4,000,040 free'd
==3335==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==3335==    by 0x4F7694C: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F78BAB: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F28DAC: Rf_mkCharLenCE (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7EC6: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:135)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335== 
==3335== Invalid read of size 1
==3335==    at 0x4F72DC6: SET_STRING_ELT (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7ED8: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:136)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4CE0F: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4712E: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4914D: ??? (in /usr/lib/R/lib/libR.so)
==3335==  Address 0x124d4043 is 3 bytes inside a block of size 4,000,040 free'd
==3335==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==3335==    by 0x4F7694C: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F78BAB: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F28DAC: Rf_mkCharLenCE (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7EC6: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:135)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335== 
==3335== Invalid write of size 8
==3335==    at 0x4F72E26: SET_STRING_ELT (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7ED8: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:136)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4CE0F: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4712E: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4914D: ??? (in /usr/lib/R/lib/libR.so)
==3335==  Address 0x127e79d8 is 3,226,008 bytes inside a block of size 4,000,040 free'd
==3335==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==3335==    by 0x4F7694C: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F78BAB: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F28DAC: Rf_mkCharLenCE (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7EC6: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:135)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335== 
==3335== Invalid read of size 1
==3335==    at 0x4F72F29: SET_VECTOR_ELT (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F6A0A: set (generic_proxy.h:83)
==3335==    by 0xD4F6A0A: operator= (generic_proxy.h:39)
==3335==    by 0xD4F6A0A: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4CE0F: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4712E: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4914D: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==  Address 0x124d4043 is 3 bytes inside a block of size 4,000,040 free'd
==3335==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==3335==    by 0x4F7694C: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F78BAB: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F28DAC: Rf_mkCharLenCE (in /usr/lib/R/lib/libR.so)
==3335==    by 0xD4F7EC6: toSEXP(feather::PrimitiveArray const*) (feather-types.cpp:135)
==3335==    by 0xD4F8957: toSEXP(std::unique_ptr<feather::Column, std::default_delete<feather::Column> > const&) (feather-types.cpp:222)
==3335==    by 0xD4F69F8: coldataFeather(Rcpp::Vector<19, Rcpp::PreserveStorage> const&, Rcpp::Vector<13, Rcpp::PreserveStorage> const&) (feather-read.cpp:102)
==3335==    by 0xD4F2AFB: feather_coldataFeather (RcppExports.cpp:50)
==3335==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3335==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3335== 

> dim(df)
[1] 500000     19
```

**Comment by dtmngngnnj** [comment]
Redoing the same but now with "--leak-check=full"
(only showing the part on exiting R) 

```
R  --debugger="valgrind --leak-check=full"
..
..
> 
Save workspace image? [y/n/c]: n
==3703== 
==3703== HEAP SUMMARY:
==3703==     in use at exit: 129,727,220 bytes in 31,523 blocks
==3703==   total heap usage: 55,214 allocs, 23,691 frees, 168,662,706 bytes allocated
==3703== 
==3703== 104 bytes in 1 blocks are possibly lost in loss record 52 of 1,926
==3703==    at 0x4C29180: operator new(unsigned long) (vg_replace_malloc.c:324)
==3703==    by 0xC670E98: std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.20)
==3703==    by 0xC672814: char* std::string::_S_construct<char const*>(char const*, char const*, std::allocator<char> const&, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.20)
==3703==    by 0xC672C45: std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.20)
==3703==    by 0xD4F4263: openFeatherTable(std::string const&) (feather-read.cpp:12)
==3703==    by 0xD4F5E68: openFeather(std::string const&) (feather-read.cpp:67)
==3703==    by 0xD4F0A9A: feather_openFeather (RcppExports.cpp:27)
==3703==    by 0x4F0BCC7: ??? (in /usr/lib/R/lib/libR.so)
==3703==    by 0x4F47547: Rf_eval (in /usr/lib/R/lib/libR.so)
==3703==    by 0x4F492C7: ??? (in /usr/lib/R/lib/libR.so)
==3703==    by 0x4F4734C: Rf_eval (in /usr/lib/R/lib/libR.so)
==3703==    by 0x4F4CE0F: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==3703== 
==3703== LEAK SUMMARY:
==3703==    definitely lost: 0 bytes in 0 blocks
==3703==    indirectly lost: 0 bytes in 0 blocks
==3703==      possibly lost: 104 bytes in 1 blocks
==3703==    still reachable: 129,727,116 bytes in 31,522 blocks
==3703==         suppressed: 0 bytes in 0 blocks
==3703== Reachable blocks (those to which a pointer was found) are not shown.
==3703== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==3703== 
==3703== For counts of detected and suppressed errors, rerun with: -v
==3703== ERROR SUMMARY: 387018 errors from 6 contexts (suppressed: 1 from 1)
```

**Comment by dtmngngnnj** [comment]
Try and see if you get the error yourself with this 8MB datafile: 
[yellow_tripdata_2015-01-500001.feather.bz2]
(http://data.munging.ninja/pubcoms/20160526_bug_feather_mem_not_mapped/yellow_tripdata_2015-01-500001.feather.bz2)

P.S.: I'll remove this file after a day or two.

**Comment by ThakurRajAnand** [comment]
I am experiencing something weird. If I write a big data frame from R to feather format then I am able to read it in same R session but if I close the session and try reading it then I am also getting same error.

I can read upto 8 GB file in same session.

**Comment by Sureya** [comment]
I have a list of 500+ file names , I am looping through to read the files in feather.

at each run , i encounter the same error stating,
```

 Error in coldataFeather(x, i) : 
  SET_STRING_ELT() can only be applied to a 'character vector', not a 'NULL' 
```

there is no pattern on which index it occurs, sometime it happens for the 15th file sometimes 93rd and so on.

But If i run the same file out of the loop it works just fine. I'm not sure why.

Any help would be really useful.

Cheers!

**Comment by wesm** [comment]
Any insights into what's going on here?

**Comment by statquant** [comment]
Hello, I am also getting this error, I am reading feather file in parallel and am getting random errors like Error in readFeather(filePath, gridTbl = dataGrid, columns) : 
  task 223 failed - "SET_STRING_ELT() can only be applied to a 'character vector', not a 'NULL'"
Usually relaunching the exact same statement works

**Comment by dtmngngnnj** [comment]
## TL;DR:

I just installed the latest version of feather, did a quick test, and then a more thorough one, and can report that **the problems is no longer**. Thanks for the fix!


## Detail:

BEFORE UPGRADE: sessionInfo() tells me version of feather is: [1] **feather_0.0.1**

Reinstall feather/R via devtools (did this on Thu Aug 11 2016, at 11h00 GMT)

    devtools::install_github("wesm/feather/R")

AFTER UPGRADE: R sessionInfo now says: [1] **feather_0.0.1.9000**

A quick, short test, proved successfull, then I reran the longer test I described here: 

http://data.munging.ninja/pubcoms/20160526_bug_feather_mem_not_mapped/

And it works fine! Great!
